### DISCUSSION TITLE
Can I define a very long rule on multiple lines?

### DISCUSSION THREAD

**Discussion by JulienGrv** [lark-parser/lark#837] [Q&A] [answered]
It's all in the title :) thanks for the answer!

**Comment by MegaIng** [comment, marked as answer]
Nothing is in the title. How does you rule look? The answer is maybe.

**Reply by JulienGrv** [reply]
I thought I'm not even going to answer considering how you responded but let's say I am on a good mood today.

Here is an example of what I mean under "very long rule":

```
module: _begin{"MODULE"} name long_identifier rule1* rule2* rule3* rule4* rule5* rule6* rule7* rule8* rule9* rule10* rule11* rule12* rule13* rule14* _end{"MODULE"}
```

I know I could split using something like:

```
module: _begin{"MODULE"} _module1 _end{"MODULE"}
_module1: name long_identifier rule1* rule2* rule3* rule4* _module2
_module2: rule5* rule6* rule7* rule8* rule9* _module3
_module3: rule10* rule11* rule12* rule13* rule14*
```

But I was thinking that there is maybe a syntax to split a rule in multiple lines. Perhaps something like:

```
module: _begin{"MODULE"} name long_identifier \
               rule1* rule2* rule3* rule4* \
               rule5* rule6* rule7* rule8* rule9* \
               rule10* rule11* rule12* rule13* rule14* _end{"MODULE"}
```

Anyway, I hope you can visualize my question now.

**Reply by MegaIng** [reply]
I put the same amount of effort into my original answer as you did into your original question. Now that you actually have a full question, I can give you a clear answer.

(there is a somewhat related discussion in the [gitter chat](https://gitter.im/lark-parser/Lobby?at=60147e5b55359c58bf0bf915). Maybe that already helps)

The answer is no you can't. A syntax like that is not supported. You can split a rule over multiple lines if you want to go with an or choice, so

```
module: rule1 | rule2 | rule3 | rule4  | ...
```

can be written as 
```
module: rule1
      | rule2
      | rule3
      | rule4
      | ...
```

This also works for nested construct, so if in your example the exact order of the rules is irrelevant, you can use

```
module: _begin{"MODULE"} name long_identifier (rule1 | rule2 | rule3 | rule4 | rule5 | rule6 | rule7 | rule8 | rule9 | rule10 | rule11 | rule12 | rule13 | rule14)* _end{"MODULE"}
```

and then
```
module: _begin{"MODULE"} name long_identifier (rule1 
      | rule2 | rule3 | rule4 | rule5 
      | rule6 | rule7 | rule8 | rule9 
      | rule10 | rule11 | rule12 
      | rule13 | rule14)* _end{"MODULE"}
```

This of course slightly changes the semantic. But it might be worth considering since normally this many rules are hard to distinguish afterwards so grouping them in some way like in you solution might be beneficial anyway.

**Reply by JulienGrv** [reply]
No offense, but I was honestly convinced my question was speaking for itself. You could have given me the benefit of the doubt and just asked for more details. Anyway, no hard feelings.

Thank you for your thorough answer! I actually think that I will use your solution because the order of the rules is irrelevant for my use case, as far as I know.
